protocol_common.go

```diff
@@ -355,6 +356,25 @@ func (h *Headscale) handleAuthKeyCommon(
 
 			return
 		}
+
+		aclTags := pak.toProto().AclTags
+		if len(aclTags) > 0 {
+			// This conditional preserves the existing behaviour, although SaaS would reset the tags on auth-key login
+			err = h.SetTags(machine, aclTags)
```

Comment:
If this _overrides_ the tags, then technically if the list is empty, we should still set it? as in, we should set it to no tags?

Reply:
So this handles some divergent behaviour between SaaS and headscale. I believe it's possible in headscale for a machine to re-auth with a pre-auth key and keep its previously set tags, whereas upstream would reset the tags in my testing.

This logic tries to keep the best of both worlds, when a machine is re-authenticated with a pak:
  * if the pak has no associated tags, the machine keeps its previous tags (the current, non SaaS behaviour)
  * if the pak has tags, these tags replace any previously set on the machine (which would more closely align with a user's *intent*)

Removing the conditional would mean that, when a machine is re-authenticated with a pak, it would lose any previous tags. This is what SaaS does but I didn't want to force a change in headscales behaviour

tl;dr I feel this is messy but its somewhat behaviour preserving

Reply:
ok, make sense, we can live with that for now, and if we decide to change it, we will have to mark it as breaking.